src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -217,12 +224,48 @@ void CreatePrimaryRequest<I>::unlink_peer() {
           // promotion
           count = 0;
           unlink_snap_id = 0;
+          prev_snap_id = 0;
           continue;
         }
         if (info->mirror_peer_uuids.count(peer) == 0) {
           // snapshot is not linked with this peer
           continue;
         }
+        if (prev_snap_id) {
+          librados::IoCtx m_group_io_ctx;
+          int r = librbd::util::create_ioctx(m_image_ctx->md_ctx,
+                                             "group", info->group_spec.pool_id,
+                                             {}, &m_group_io_ctx);
+          if (r < 0) {
+            return;
+          }
+          cls::rbd::GroupSnapshot prev_group_snap;
+          std::string group_header_oid = librbd::util::group_header_name(
+                                                 info->group_spec.group_id);
+          r = cls_client::group_snap_get_by_id(&m_group_io_ctx,
+                                               group_header_oid,
+                                               prev_group_snap_id,
+                                               &prev_group_snap);
+          if (r < 0) {
+            lderr(cct) << "failed to retrieve group snapshot: "
+                       << cpp_strerror(r) << dendl;
+            prev_snap_id = snap_info_pair.first;
+            prev_group_snap_id = info->group_snap_id;
+            continue;
+          }
+          if (prev_group_snap.state != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+            peer_uuid = peer;
```

Comment:
This should not be done in this class as an image level snapshot operation should not be trigering  groupsnap unlink ops.. A separate group snapshot class should handle this.